Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  /// which is consistent with C# IIUC. I need to perform the experiement to verify this.
+  ///
+  /// REVIEW NOTE: I'm going with component as return type, which implies the more common
+  /// use would be to read or inspect the root, rather than forming new paths from the same root.
+  /// For that use case, you can do `FilePath(otherPath.root)`, or use
+  /// the `relativePath`'s setter, or use the `/` operator (if we go that route).
+  public var root: FilePath.Component? {
+    get {
+      let rootEnd = _parseRoot().0
+      guard rootEnd != storage.startIndex else { return nil }
+      return Component(self, ..<rootEnd)
+    }
+    set {
+      fatalError()
+    }
+  }
+
+  /// REVIEW NOTE: I'm dropping rootName and rootDirectory. I think
+  /// C++ was a little misguided here after looking more at C#. I think
+  /// a better alternative would be to actually identify and work with
+  /// the root when on Windows, more akin to Rust's `Prefix` type.
+
+  /// Gets or sets the relative portion of the path (everything after root)
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/foo/bar => foo/bar`
+  ///   * `foo/bar => foo/bar`
+  ///   * `/ => ""`
+  /// * Windows:
+  ///   * `C:\foo\bar => foo\bar`
+  ///   * `foo\bar => foo\bar`
+  ///   * `\\?\UNC\server\share\file => file`
+  ///   * `\ => ""`
+  ///
+  /// REVIEW NOTE: Not optional, can return empty paths. Empty paths
+  /// are relative (but are not roots), hence non-optional.
+  public var relativePath: FilePath {
+    get {
+      let (_, relIdx) = _parseRoot()
+      return FilePath(slice: storage[relIdx...])
+    }
+    set { fatalError() }
+  }
+
+  /// Returns the final component of the path. If the path is a normal file, this
+  /// will be the file name. If the path is a directory, this is the directory name.
+  /// Returns `nil` if the path is empty or only contains a root
+  ///
+  /// REVIEW NOTE: We follow  `basename` semantics, unlike `file_name` in
+  /// C++ which treats trailing `/` specially, or in Rust which returns nil for `..`
+  ///
+  /// REVIEW NOTE: shell and C basename will return `/` if the path is `/`. I don't
+  /// think we want to, it seems clearer that this only is in the relative portion of the path,
+  /// and doing so would make the setter weird.
+  public var basename: FilePath.Component? {
+    get {
+      let (rootEnd, _) = _parseRoot()
+      guard let base = self.components.last, base.slice.startIndex >= rootEnd else {
+        return nil
+      }
+      return base
+    }
+    set {
+      // TODO: this is assuming that pop correctly preserves root
+      _ = self.pop()
+      if let c = newValue { self.append(FilePath(c)) }
+    }
+  }
+
+  /// Returns the path up to but not including the `basename`.
+  ///
+  /// If the path only contains a root, returns `self`.
+  /// If the path has no root and only includes a single component,
+  /// returns an empty FilePath.
```

Comment:
This seems like the right choice to me, but then I would expect `basename` to also return an empty string rather than optional.